### PR TITLE
Add more documentations to DefaultIndexMap and PalDBIndexMap, and fix…

### PR DESCRIPTION
**Pull request by XianXing**
… some related Scala style issues.

### CODE REVIEW

File: photon-ml/src/main/scala/com/linkedin/photon/ml/util/DefaultIndexMap.scala [hunk 1]
```diff
@@ -36,7 +41,9 @@ class DefaultIndexMap(@transient val featureNameToIdMap: Map[String, Int]) exten
     _idToNameMap.get(idx)
   }
 
-  override def getIndex(name: String): Int = featureNameToIdMap.getOrElse(name, IndexMap.NULL_KEY)
+  override def getIndex(name: String): Int = {
+    featureNameToIdMap.getOrElse(name, IndexMap.NULL_KEY)
+  }
```
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/util/DefaultIndexMap.scala]: The previous way is correct according to the Scala style guide ([here specifically](http://docs.scala-lang.org/style/declarations.html#body)).

File: photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala [hunk 2]
```diff
@@ -157,7 +158,8 @@ class PalDBIndexMap extends IndexMap {
 
   //noinspection ScalaStyle
   override def -(key: String): Map[String, Int] = {
-    throw new RuntimeException("This map is a read-only immutable map, remove operation is unsupported.")
+    throw new RuntimeException("This map is a read-only immutable map, " +
+      "remove operation is unsupported.")
```
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala]: Any reason for this change and the one above it? The exception messages fit on one line.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala [hunk 3]
```diff
@@ -173,12 +175,18 @@ object PalDBIndexMap {
...
-    s"paldb-partition-${namespace}-${partitionId}.dat"
+  def partitionFilename(
+    partitionId: Int,
+    namespace: String = IndexMap.GLOBAL_NS): String = {
 
-  class PalDBIndexMapIterator(private val indexMap: PalDBIndexMap) extends Iterator[(String, Int)] {
-    private var _iter: Iterator[JMap.Entry[Any, Any]] = null
-    private var _currentItem: (String, Int) = null
+    s"paldb-partition-$namespace-$partitionId.dat"
+  }
```
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala]: The previous way is correct according to the Scala style guide ([here specifically](http://docs.scala-lang.org/style/declarations.html#body)).

File: photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala (hunk 2 above)
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala]: It appears to my IDE that it does not fit in one line (length of this line == 106).

File: photon-ml/src/main/scala/com/linkedin/photon/ml/util/DefaultIndexMap.scala (hunk 1 above)
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/util/DefaultIndexMap.scala]: It appears to my IDE that it does not fit in one line (length of this line == 99).

File: photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala (hunk 3 above)
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala]: It appears to my IDE that it does not fit in one line (length of this line == 101).

File: photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala [hunk 4]
```diff
@@ -198,19 +202,22 @@ object PalDBIndexMap {
...
       }
 
       while (_iter.hasNext && _currentItem == null) {
         val entry = _iter.next()
-        if (entry.getKey().isInstanceOf[String]) {
-          val key = entry.getKey().asInstanceOf[String]
-          val i = indexMap.getPartitionId(key)
-          val value = entry.getValue().asInstanceOf[Int] + indexMap._offsets(i)
-          _currentItem = (key, value)
+        entry.getKey match {
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala]: Nice! We could also match `entry.getValue` here and eliminate the other `asInstanceOf`, e.g.:

```scala
(entry.getKey, entry.getValue) match {
  case (key: String, entryValue: Int) =>
     val i = indexMap.getPartitionId(key)
     val value = entryValue + indexMap._offsets(i)
     _currentItem = (key, value)
  case _ =>
}
```

Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/util/PalDBIndexMap.scala]: Great! I really missing having a Scala expert around!

**Pull request merged**